File: frontend/lib/api/handoff.ts
import { API_URL, ApiError, API_FETCH_TIMEOUT_MS, apiFetch } from "@/lib/api/client";

export type HandoffPurpose = "answer" | "feedback";

export interface HandoffCreateResponse {
  token: string;
  capture_url: string;
  expires_at: string;
}

export interface CaptureMetaResponse {
  purpose?: HandoffPurpose;
  session_id?: string | null;
  homework_id?: string | null;
  position?: number | null;
  task_title: string | null;
  question_preview: string | null;
  expires_at: string;
  already_has_photo: boolean;
  staged_image_id?: string | null;
  staged_image_url?: string | null;
}

export interface CaptureUploadResponse {
  purpose?: HandoffPurpose;
  position?: number | null;
  answer_image_ids?: string[];
  answer_image_urls?: string[];
  staged_image_id?: string | null;
  staged_image_url?: string | null;
}

export function createHandoff(
  sessionId: string,
  position: number,
): Promise<HandoffCreateResponse> {
  return apiFetch<HandoffCreateResponse>(
    `/api/tests/sessions/${sessionId}/steps/${position}/handoff`,
    { method: "POST" },
  );
}

export function createFeedbackHandoff(
  assignmentId: string,
  position?: number | null,
): Promise<HandoffCreateResponse> {
  return apiFetch<HandoffCreateResponse>(
    `/api/homework/${assignmentId}/feedback-handoff`,
    {
      method: "POST",
      body: JSON.stringify(
        position === undefined ? {} : { position },
      ),
    },
  );
}

export function getCaptureMeta(token: string): Promise<CaptureMetaResponse> {
  return apiFetch<CaptureMetaResponse>(`/api/capture/${token}`);
}

export async function captureUpload(
  token: string,
  file: File,
): Promise<CaptureUploadResponse> {
  const formData = new FormData();
  formData.append("file", file);

  const useTimeout = process.env.VITEST !== "true";
  const response = await fetch(`${API_URL}/api/capture/${token}`, {
    method: "POST",
    credentials: "include",
    body: formData,
    ...(useTimeout ? { signal: AbortSignal.timeout(API_FETCH_TIMEOUT_MS) } : {}),
  });

  if (!response.ok) {
    let detail = response.statusText;
    try {
      const data = await response.json();
      if (typeof data?.detail === "string") {
        detail = data.detail;
      }
    } catch {
      // keep status text
    }
    throw new ApiError(response.status, detail);
  }

  return (await response.json()) as CaptureUploadResponse;
}

File: backend/app/schemas/handoff.py
"""Schemas for QR handoff / mobile capture (SPEC §1.9.9)."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class HandoffCreateResponse(BaseModel):
    token: uuid.UUID
    capture_url: str
    expires_at: datetime


class FeedbackHandoffCreate(BaseModel):
    position: int | None = None


class CaptureMetaResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    purpose: Literal["answer", "feedback"] = "answer"
    session_id: uuid.UUID | None = None
    homework_id: uuid.UUID | None = None
    position: int | None = None
    task_title: str | None = None
    question_preview: str | None = None
    expires_at: datetime
    already_has_photo: bool
    staged_image_id: uuid.UUID | None = None
    staged_image_url: str | None = None


class CaptureUploadResponse(BaseModel):
    purpose: Literal["answer", "feedback"] = "answer"
    position: int | None = None
    answer_image_ids: list[uuid.UUID] = Field(default_factory=list)
    answer_image_urls: list[str] = Field(default_factory=list)
    staged_image_id: uuid.UUID | None = None
    staged_image_url: str | None = None
